src/main/scala/org/phenoscape/kb/PresenceAbsenceOfStructure.scala

```diff
@@ -151,28 +163,81 @@ object PresenceAbsenceOfStructure {
+        s"Generated from the Phenoscape Knowledgebase on $date by Ontotrace query" // :\n* taxa: ${taxonClass.asOMN.getLiteralLexicalForm}\n* entities: ${entityClass.asOMN.getLiteralLexicalForm}\n* variable only: ${variableOnly}"
       )
       dataset
     }
   }
 
   private def unOBO(uri: String): String = uri.replaceAllLiterally("http://purl.obolibrary.org/obo/", "")
 
-  def buildMatrixQuery(entityClass: OWLClassExpression, taxonClass: OWLClassExpression): Query =
-    construct(t('taxon, 'relation, 'entity),
-              t('taxon, rdfsLabel, 'taxon_label),
-              t('entity, rdfsLabel, 'entity_label)) from "http://kb.phenoscape.org/" where (bgp(
-      t('taxon, 'relation, 'entity),
-      t('taxon, rdfsLabel, 'taxon_label),
-      t('entity, rdfsLabel, 'entity_label),
-      t('entity, rdfsSubClassOf, entityClass.asOMN),
-      t('taxon, rdfsSubClassOf, taxonClass.asOMN)
-    ),
-    new ElementFilter(
-      new E_OneOf(new ExprVar('relation),
-                  new ExprList(
-                    List[Expr](new NodeValueNode(has_presence_of), new NodeValueNode(has_absence_of)).asJava))))
+  def buildMatrixQuery(entityClass: Option[OWLClassExpression],
+                       entityList: Option[Seq[IRI]],
+                       taxonClass: Option[OWLClassExpression],
+                       taxonList: Option[Seq[IRI]],
+                       includeSubClasses: Boolean): Query = {
+    val entityClassExprSubClasses = sparql""" ?entity $rdfsSubClassOf ${entityClass.get.asOMN} . """
+    val entityClassExprQuery = {
+      sparql"""
+               ?entity ${rdfType} ${entityClass.get.asOMN} .
+
+              """
+    } + {
+      if (includeSubClasses) sparql"$entityClassExprSubClasses" else sparql""""""
+    }
+    val entityListStr = entityList.mkString(" ")
+    val entityListQuery =
+      sparql"""
+                VALUES ?entity {${entityListStr}}
+              """
+    val entityQuery =
+      if (entityClass.isDefined && entityList.isDefined) sparql"$entityClassExprQuery" + sparql"$entityListQuery"
+      else if (entityClass.isDefined) entityClassExprQuery
+      else entityListQuery
+
+    val taxonClassExprSubClasses =
+      sparql"""
+               ?taxon $rdfsSubClassOf ${taxonClass.get.asOMN} .
+              """
+    val taxonClassExprQuery = {
+      sparql"""
+               ?taxon $rdfType ${taxonClass.get.asOMN} .
+              """
+    } + {
+      if (includeSubClasses) sparql"$taxonClassExprSubClasses" else sparql""""""
+    }
+    val taxonListStr = taxonList.mkString(" ")
+    val taxonListQuery =
+      sparql"""
+                VALUES ?taxon {${taxonListStr}}
+              """
+    val taxonQuery =
+      if (taxonClass.isDefined && taxonList.isDefined) sparql"$taxonClassExprQuery" + sparql"$taxonListQuery"
```

Comment:
Much cleaner to use a for comprehension than `Option.isDefined`:

```scala
for {
  taxonCls <- taxonClass
  taxa <- taxonList
} yield {
// some stuff using both
}
```

But this would be cleaner just to create the SPARQL text separately for each and include each in the main query (either could end up being an empty string).